src/python_testing/TC_ECOINFO_2_1.py

```diff
@@ -0,0 +1,183 @@
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+import chip.clusters as Clusters
+from chip.clusters.Types import NullValue
+from chip.interaction_model import Status
+from chip.tlv import uint
+from matter_testing_support import MatterBaseTest, TestStep, async_test_body, type_matches
+from mobly import asserts
+
+
+class TC_ECOINFO_2_1(MatterBaseTest):
+
+    def steps_TC_ECOINFO_2_1(self) -> list[TestStep]:
+        steps = [TestStep(1, "Identify endpoints with Ecosystem Information Cluster", is_commissioning=True),
+                 TestStep(2, "Reading RemovedOn Attribute"),
+                 TestStep(3, "Reading DeviceDirectory Attribute"),
+                 TestStep(4, "Reading LocationDirectory Attribute"),
+                 TestStep(5, "Try Writing to RemovedOn Attribute"),
+                 TestStep(6, "Try Writing to DeviceDirectory Attribute"),
+                 TestStep(7, "Try Writing to LocationDirectory Attribute"),
+                 TestStep(8, "Repeating steps 2 to 7 for each endpoint identified in step 1")]
+        return steps
+
+    @async_test_body
+    async def test_TC_ECOINFO_2_1(self):
+        dev_ctrl = self.default_controller
+        dut_node_id = self.dut_node_id
+
+        self.print_step(0, "Commissioning, already done")
+
+        first_cluster_validation = True
+        self.step(1)
+        endpoint_wild_card_read = await dev_ctrl.ReadAttribute(dut_node_id, [(Clusters.EcosystemInformation.Attributes.ClusterRevision)])
+        list_of_endpoints = list(endpoint_wild_card_read.keys())
+
+        for cluster_endpoint in list_of_endpoints:
+            if first_cluster_validation:
+                self.step(2)
+            removed_on = await self.read_single_attribute(
+                dev_ctrl,
+                dut_node_id,
+                endpoint=cluster_endpoint,
+                attribute=Clusters.EcosystemInformation.Attributes.RemovedOn)
+
+            removed_on_is_null = removed_on is NullValue
+            if not removed_on_is_null:
+                asserts.assert_true(type_matches(removed_on, uint))
+                asserts.assert_greater(removed_on, 0, "RemovedOn must be greater than 0", "RemovedOn should be a uint")
+
+            if first_cluster_validation:
+                self.step(3)
+            device_directory = await self.read_single_attribute(
+                dev_ctrl,
+                dut_node_id,
+                endpoint=cluster_endpoint,
+                attribute=Clusters.EcosystemInformation.Attributes.DeviceDirectory,
+                fabricFiltered=False)
+
+            num_of_devices = len(device_directory)
+            if removed_on_is_null:
```

Comment:
would it make sense to split out the removed_on validation in a separate method? The indents here get quite large